In [1]:
import pandas as pd
import numpy as np
import datetime
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping EmployeeData (HVC_AM0)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#Reading the file and storing each line in the list text
with open ("HVC_AM0.csv", "r") as f:
    textEmployees = f.readlines()
textEmployees.pop(0)

#We make a dictionnary based on the data
CompanyEmployeesDict = {}

for e in textEmployees:
    Employee_ID, EmployeeNumber, Name, Postcode, Location, Language = e.replace("\n","").split(";")
    Employee_ID=int(Employee_ID)
    Postcode=int(Postcode)
    CompanyEmployeesDict[Employee_ID]= {"EmployeeNumber":EmployeeNumber,"Name": Name, "Postcode": Postcode, "Location": Location,"Language": Language }


#Some Employees do not have a EmployeeNumber: We set the EmployeeNumber of these employees to 0
#In this way we can convert all to a float
for e in CompanyEmployeesDict:
    E_ID = CompanyEmployeesDict[e]["EmployeeNumber"]
    if(E_ID == ""):
        CompanyEmployeesDict[e]["EmployeeNumber"]="0"
    CompanyEmployeesDict[e]["EmployeeNumber"] = int(float(CompanyEmployeesDict[e]["EmployeeNumber"]))   

df_CompanyEmployees = pd.DataFrame.from_dict(CompanyEmployeesDict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping ProductData (HVC_AR0)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_AR0.csv", "r") as f:
    textProducts = f.readlines()
textProducts.pop(0)

CompanyProductsDict = {}
for e in textProducts:
    ProductID, Description, Category, Family, Price = e.replace("\n","").split(";")
    ProductID = int(ProductID)
    CompanyProductsDict[ProductID]= {"Description":Description,"Category": Category, "Family": Family, "Price": Price}
   
#Some prices are missing   
for e in CompanyProductsDict:
  PriceCheck = CompanyProductsDict[e]["Price"]
  if(PriceCheck==""):
        CompanyProductsDict[e]["Price"]="0.0"
  CompanyProductsDict[e]["Price"]=float(CompanyProductsDict[e]["Price"])  

#Whenever they are Coupons they are not assigned a family
#We assign Coupon to its Family for clearity
for e in CompanyProductsDict:
  PriceCheck = CompanyProductsDict[e]["Family"]
  if(PriceCheck==""):
        CompanyProductsDict[e]["Family"]="Coupon"

df_Products = pd.DataFrame.from_dict(CompanyProductsDict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping Reviews Data (HVC_CUSTOMER_REVIEWS)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------   


with open ("HVC_CUSTOMER_REVIEWS.csv", "r") as f:
    textCustReviews = f.readlines()
textCustReviews.pop(0)

CompanyReviewsDict = {}
for e in textCustReviews:
    CustomerID, Review  = e.replace("\n","").split(";")
    CustomerID = int(CustomerID)
    CompanyReviewsDict[CustomerID]= {"Review": Review}

df_Reviews = pd.DataFrame.from_dict(CompanyReviewsDict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping Depot data (HVC_DEPOT)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#Reading the file and storing each line in the list text
with open ("HVC_DEPOT.csv", "r") as f:
    text_Depot = f.readlines()
text_Depot.pop(0)

#We make a dictionnary based on the data
Depot_Dict = {}

for dep in text_Depot:
    Depot_Id, Depot_Location, Depot_Route = dep.replace("\n","").split(";")
    Depot_Id = int(Depot_Id)
    Depot_Route = int(Depot_Route)
    Depot_Dict[Depot_Id] = {"DEPOT":Depot_Location, "HVROUTETEMPLATE_NRID":Depot_Route}

#The Depot_Id and HVROUTETEMPLATE_NRID are both integers, Depot_Location
#The names of the keys are based on the table descriptions for easy looking up in case of further calculations
#There are no missing values in our dictionnary so further preprocessing is not necessary.

df_Depot = pd.DataFrame.from_dict(Depot_Dict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping DaysOfWeek (HVC_HVDAYOFWEEK)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_HVDAYOFWEEK.csv", "r") as f:
    text_DayOfWeek = f.readlines()
text_DayOfWeek.pop(0)

#We make a dictionnary based on the data
DayOfWeek_Dict = {}

for day in text_DayOfWeek:
    HVDAYOFWEEK_NRID, DAY = day.replace("\n","").split(";")
    HVDAYOFWEEK_NRID = int(float(HVDAYOFWEEK_NRID))
    DayOfWeek_Dict[int(HVDAYOFWEEK_NRID)] = {"DAY":DAY, "HVDAYOFWEEK_NRID":HVDAYOFWEEK_NRID}


#The HVDAYOFWEEK_NRID is an integer, DAY is a string. 
#!! important notice for further caclulations, the first day (day with id = 1) is a sunday! Not a monday!
df_DaysOfWeek = pd.DataFrame.from_dict(DayOfWeek_Dict, orient = 'index')



#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping Position data (HVC_HVPOSITION)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_HVPOSITION.csv", "r") as f:
    text_Position = f.readlines()
text_Position.pop(0)

#We make a dictionnary based on the data
Position_Dict = {}

for loc in text_Position:
    SO0_NRID, LONG, LAT = loc.replace("\n","").split(";")
    SO0_NRID = int(float(SO0_NRID))
    LONG = float(LONG)
    LAT = float(LAT)
    Position_Dict[SO0_NRID] = {"LONG":LONG, "LAT":LAT}

#SO0_NRID is int, LONG and LAT are both float values.

df_Position = pd.DataFrame.from_dict(Position_Dict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping Route Template data (HVC_HVROUTETEMPLATE)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_HVROUTETEMPLATE.csv", "r") as f:
    textRouteTemplate = f.readlines()
    textRouteTemplate.pop(0)
    
RouteTemplate_Dict = {}

for route in textRouteTemplate:
    HVROUTETEMPLATE_NRID, REGION, WEEKORDER, HVDAYOFWEEK_NRID = route.replace("\n","").split(";")
    RouteTemplate_Dict[int(float(HVROUTETEMPLATE_NRID))] = {"REGION": REGION, "WEEKORDER": int(float(WEEKORDER)), 
                                                "DAYOFWEEK_NRID": int(float(HVDAYOFWEEK_NRID))}        

df_RouteTemplate = pd.DataFrame.from_dict(RouteTemplate_Dict, orient = 'index')
                                         


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping Visit Outcome data (HVC_HVVISITOUTCOME)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_HVVISITOUTCOME.csv", "r") as f:
    textVisitOutcome = list()
    for line in f:
        line = line.replace("\n","")
        obs = line.split(";")
        textVisitOutcome.append(obs)

column_names = textVisitOutcome[0]
textVisitOutcome.pop(0)
VisitOucome_Dict = {}

for visit in textVisitOutcome:
    VisitOucome_Dict[visit[1]] = visit[0]


df_VisitOutcome = pd.DataFrame.from_dict(VisitOucome_Dict, orient = 'index')


#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping TransactionData (HVC_HVVISITRESULT)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#Reading the file and storing each line in the list text
with open ("HVC_HVVISITRESULT.csv", "r") as f:
    textVisit = f.readlines()
textVisit.pop(0)

#We make a dictionnary based on the data
CompanyVisitDict = {}

for e in textVisit:
    Visit_ID, Customer_ID, Employee_ID, VisitOutcome_ID, Time, Date, Amount, paymentMethod = e.replace("\n","").split(";")
    
    Visit_ID, Customer_ID, Employee_ID, VisitOutcome_ID, Time, Date, Amount, paymentMethod = e.replace("\n","").split(";")
    Visit_ID= int(Visit_ID)

    yYear, mMonth, dDay = Date.split("-")
    yYear = int(yYear)
    mMonth = int(mMonth)
    dDay = int(dDay)
    Date=datetime.date(yYear,mMonth,dDay)

    Customer_ID= int(Customer_ID)
    Employee_ID= int(Employee_ID)
    VisitOutcome_ID= int(VisitOutcome_ID)
    Amount= float(Amount)
    CompanyVisitDict[Visit_ID]= {"Customer_ID":Customer_ID,"Employee_ID": Employee_ID, "VisitOutcome_ID": VisitOutcome_ID, "Time": Time,"Date": Date, "Amount": Amount, "paymentMethod": paymentMethod }

#Some visits do not have a payment method recorded: We set the paymentMethod of these cases to the None variable
#Maybe these missing values will have value later on
for e in CompanyVisitDict:
    P_method = CompanyVisitDict[e]["paymentMethod"]
    if(P_method == ""):
        CompanyVisitDict[e]["paymentMethod"]= None

df_CompanyVisit = pd.DataFrame.from_dict(CompanyVisitDict, orient = 'index')

#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping EmployeeData (HVC_SO0)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#Reading the file and storing each line in the list text
with open ("HVC_SO0.csv", "r") as f:
    textCustomerType = f.readlines()
textCustomerType.pop(0)

#We make a dictionnary based on the data
CustomerTypeDict = {}

for e in textCustomerType:
    Customer_ID, RouteTemplate_ID, CustomerType, Postcode, Language, Season = e.replace("\n","").split(";")
    Customer_ID=int(float(Customer_ID))
    RouteTemplate_ID=int(float(RouteTemplate_ID))
    CustomerTypeDict[Customer_ID]= {"HVROUTETEMPLATE_NRID":RouteTemplate_ID,"CustomerType": CustomerType, "Postcode": Postcode, "Language": Language,"Season": Season}


df_CustomerType = pd.DataFrame.from_dict(CustomerTypeDict, orient = 'index')

#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#Prepping ProductData (HVC_VISITRESULTDETAILS)
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


with open ("HVC_VISITRESULTDETAILS.csv", "r") as f:
    textVisitDetails = f.readlines()
textVisitDetails.pop(0)


#We make a dictionnary based on the data
VisitDetailsDict = {}


for e in textVisitDetails:
    VisitDetails_ID, Product_ID, Quantity, Visit_ID = e.replace("\n","").split(";")
    VisitDetails_ID=int(VisitDetails_ID)
    Product_ID=int(Product_ID)
    Quantity=int(float(Quantity))
    Visit_ID=int(float(Visit_ID))
    VisitDetailsDict[VisitDetails_ID]= {"Product_ID":Product_ID,"Quantity": Quantity, "Visit_ID": Visit_ID}

df_VisitDetails = pd.DataFrame.from_dict(VisitDetailsDict, orient = 'index')


#---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#This gives us the following 11 dataframes to work with:
# 1  df_CompanyEmployees (HVC_AM0)
# 2  df_Products (HVC_AR0)
# 3  df_Reviews  (HVC_CUSTOMER_REVIEWS)
# 4  df_Depot  (HVC_DEPOT)
# 5  df_DaysOfWeek (HVC_HVDAYOFWEEK)
# 6  df_Position (HVC_HVPOSiTION)
# 7  df_RouteTemplate (HVC_HVROUTETEMPLATE)
# 8  df_VisitOutcome (HVC_HVVISITOUTCOME)
# 9  df_CompanyVisit (HVC_VISITRESULT)
# 10 df_CustomerType (HVC_SO0)
# 11 df_VisitDetails (HVC_VISITRESULTDETAILS)
#---------------------------------------------------------------------------------------------------------------------------------------------------------------------------



In [40]:
#I imported datetime
#HVROUTETEMPLATE weekorder and dayofweek are no integers, I changed that.
#added a name to the columns for daysofweek in the daysofweek database
#df_CustomerType changed the name of the final dictionairy because it created a new dictionair that didnt make sense
#changed the id for routetemplate to an integer
#for HVC_SO0 i changed the columname to HVROUTETEMPLATE_NR for the route ID
#there are no column names for visitoutcome, also the id needs to be an integer.
#all overlapping colmuns should have the same name, is this the case? maybe take over the name from the table document? 


In [43]:
df_CompanyEmployees.head(5)

EmployeeNumber            Name  Postcode  Location Language
18503611215721            1001  BOONE Frederik      2460    Tielen    nl-BE
18503612347289            1021    JANSSENS Jos      2400       Mol    nl-BE
18503612452173            1022  LEYSEN Michiel      3500   Hasselt    nl-BE
18503621528962            1122   ALLEMAN Benny      3900  Overpelt    nl-BE
18503622652855            1135   GOOVAERTS Jef      3920    Lommel    nl-BE

In [25]:
df_Products.head(5)

Description         Category              Family  \
20742288286838        Dessert Caramel/14  Verkoopsartikel                Cups   
20413488326838                Penalty/16  Verkoopsartikel            Hand Ice   
21515648414638              Kriekentaart  Verkoopsartikel              Coffee   
20296288686840  Expo met zachte wafel/12  Verkoopsartikel  Individual Dessert   
21844652504048                Paaseitjes  Verkoopsartikel  Individual Dessert   

                Price  
20742288286838   10.9  
20413488326838    6.5  
21515648414638    7.9  
20296288686840    7.3  
21844652504048    7.5

In [24]:
df_Reviews.head(5)

Review
687782  Good service, satisfied
718230  Good service, satisfied
718236  Good service, satisfied
718247               No opinion
718257  Good service, satisfied

In [71]:
df_Depot

DEPOT  HVROUTETEMPLATE_NRID
0      Deinze             219020601
1      Deinze             219021241
2      Deinze             219024456
3      Deinze             219021882
4      Deinze             219022524
5      Deinze             219023167
6      Deinze             219023811
7      Deinze             219019324
8      Deinze             219019962
9      Deinze             289414945
10     Deinze             289757472
11     Deinze              70671404
12  Antwerpen             219018687
13  Antwerpen             289756470
14  Antwerpen             289755468
15  Antwerpen             289759476
16  Antwerpen             289763484
17  Antwerpen             289758474
18  Antwerpen             289762482
19  Antwerpen             289764486
20  Antwerpen             289760478
21  Vilvoorde             289416949
22  Vilvoorde             289761480
23  Vilvoorde             289419955
24  Vilvoorde             289417951
25  Vilvoorde             289418953
26  Vilvoorde             289411939
27  Vilvoorde             289412941
28  Vilvoorde             289661606
29  Vilvoorde             289413943
30  Vilvoorde             289655594
31  Vilvoorde             289659602
32  Vilvoorde             289660604
33  Vilvoorde             289656596
34  Vilvoorde             289656596
35  Vilvoorde             289664612
36   Aarschot             289663610
37   Aarschot             289662608
38   Aarschot             289658600
39   Aarschot             289657598

In [52]:
df_DaysOfWeek.head(5)

DAY  HVDAYOFWEEK_NRID
1     Sunday                 1
2     Monday                 2
3    Tuesday                 3
4  Wednesday                 4
5   Thursday                 5

In [26]:
df_Position.head(5)

LONG        LAT
724196  3.412928  50.910235
724219  3.409088  50.903873
724231  3.407691  50.901559
724236  3.405464  50.898224
724241  3.404218  50.896691

In [15]:
df_RouteTemplate

REGION  WEEKORDER  DAYOFWEEK_NRID  HVROUTETEMPLATE_NRID
70671404     C04          1               1              70671404
219018687    C04          1               2             219018687
219019324    C04          1               3             219019324
219019962    C04          1               4             219019962
219020601    C04          1               5             219020601
219021241    C04          1               6             219021241
219021882    C04          2               2             219021882
219022524    C04          2               3             219022524
219023167    C04          2               4             219023167
219023811    C04          2               5             219023811
219024456    C04          2               6             219024456
289411939    B02          1               3             289411939
289412941    B02          1               4             289412941
289413943    B02          1               5             289413943
289414945    B02          1               6             289414945
289416949    B02          2               3             289416949
289417951    B02          2               4             289417951
289418953    B02          2               5             289418953
289419955    B02          2               6             289419955
289655594    C17          1               2             289655594
289656596    C17          1               3             289656596
289657598    C17          1               4             289657598
289658600    C17          1               5             289658600
289659602    C17          1               6             289659602
289660604    C17          2               2             289660604
289661606    C17          2               3             289661606
289662608    C17          2               4             289662608
289663610    C17          2               5             289663610
289664612    C17          2               6             289664612
289755468    B25          1               2             289755468
289756470    B25          1               3             289756470
289757472    B25          1               4             289757472
289758474    B25          1               5             289758474
289759476    B25          1               6             289759476
289760478    B25          2               2             289760478
289761480    B25          2               3             289761480
289762482    B25          2               4             289762482
289763484    B25          2               5             289763484
289764486    B25          2               6             289764486

In [77]:
len(df_RouteTemplate.index)

39

In [4]:
df_VisitOutcome.head(5)

0
NOTHING NEEDED  0.0
NOT HOME        1.0
SALES           2.0

In [29]:
df_CompanyVisit.head(5)

Customer_ID     Employee_ID  VisitOutcome_ID                Time  \
55681456       719952  23186288605042                2  12:20:01,000000000   
55681507      1453609  19946584592834                2  12:13:50,000000000   
55681509      1453614  19946584592834                2  12:14:08,000000000   
55681512      1453602  19946584592834                1  12:16:04,000000000   
55674074      1252497  18934048107254                2  10:41:09,000000000   

                Date  Amount paymentMethod  
55681456  2019-05-06    16.9          cash  
55681507  2019-05-06     7.4          cash  
55681509  2019-05-06    18.9          cash  
55681512  2019-05-06     0.0          None  
55674074  2019-05-06    34.9           tbp

In [72]:
df_CustomerType

HVROUTETEMPLATE_NRID CustomerType Postcode Language    Season
721104                     219020601      Private     3930    nl-BE  All time
721110                     219020601      Private     3930    nl-BE  All time
721121                     219020601      Private     3930    nl-BE  All time
721126                     219020601      Private     3930    nl-BE  All time
721131                     219020601      Private     3930    nl-BE  All time
...                              ...          ...      ...      ...       ...
26651656233234             219020601       Horeca     3930    nl-BE  All time
26707966298860             289760478       Horeca     3650    nl-BE  All time
26869549647542             289760478      Private     3680    nl-BE  All time
26959353239340             289418953      Private     3550    nl-BE  All time
26961148104438             219024456      Private     3950    nl-BE  All time

[5709 rows x 5 columns]

In [33]:
df_VisitDetails.head(5)

Product_ID  Quantity  Visit_ID
56055671  17545160268040         1  56055665
56055672           23699         2  56055665
56055675           23721         1  56055674
56010656           23714         1  56010655
56009821           23716         1  56009820

In [59]:
# -> 2.1. What do customers buy?
CompanyTotalProductSales = {}
Aureum = df_VisitDetails.groupby("Product_ID")["Quantity"].sum()
newAU = pd.merge(Aureum, df_Products, how="left", on=Aureum.keys())
Revenue = newAU["Price"]*newAU["Quantity"]
newAU["RevenueProduct"] = Revenue

#Best products based on revenue
BestRevProducts = newAU.sort_values(by="RevenueProduct", ascending=False)

#Most products sold in ammount
BestAmountSoldProducts = newAU.sort_values(by="Quantity", ascending=False)

#Based on Family
#Generated Revenue
BestRevFamily=BestRevProducts.groupby("Family")["RevenueProduct"].sum().sort_values(ascending=False)

#Based on Family
#Amount sold
BestAmountFamily=BestAmountSoldProducts.groupby("Family")["Quantity"].sum().sort_values(ascending=False)

#Based on Catogery
#Generated Revenue
BestRevCat=BestRevProducts.groupby("Category")["RevenueProduct"].sum().sort_values(ascending=False)

#Based on Catogery
#Amount sold
BestAmountCat=BestRevProducts.groupby("Category")["Quantity"].sum().sort_values(ascending=False)


In [ ]:
#-> 2.6. Do customers have different buying patterns during the weekend?
#This method uses the weekday function to determine the day of the week starting 
#At 0 for Monday and ending on 6 for Sunday for a given month and year
#In this way if the returned number is larger than 4 (Friday) it is a weekendDay
def getWeekIndex(d: datetime.date):
    if(d.weekday()>4):
        return "weekend"
    else:
        return "weekday"

#We add a new Column explaining wheter it is a weekday or weekend based on the Date index
df_CompanyVisit["DateIndex"] = df_CompanyVisit["Date"].apply(getWeekIndex)
df_CompanyVisit["Visit_ID"] = df_CompanyVisit.index

CompleteMerge = pd.merge(df_VisitDetails, df_CompanyVisit, on="Visit_ID")


df_Products["Product_ID"] = df_Products.index
CompleteMergeProd = pd.merge(CompleteMerge,df_Products, on="Product_ID").drop(columns=["Language", "paymentMethod", "Amount", "Time"])
CompleteMergeProd["Revenue"]=CompleteMergeProd["Quantity"]*CompleteMergeProd["Price"]

OnlyWeekdays= CompleteMergeProd[CompleteMergeProd["DateIndex"] != "weekend"]
OnlyWeekends= CompleteMergeProd[CompleteMergeProd["DateIndex"] != "weekday"]

OnlyWeekdaysCus = OnlyWeekdays.groupby(["Customer_ID","DateIndex"])["Revenue"].sum()
OnlyWeekendsCus = OnlyWeekends.groupby(["Customer_ID","DateIndex"])["Revenue"].sum()
Comparison = pd.merge(OnlyWeekendsCus,OnlyWeekdaysCus, on="Customer_ID")

#X is weekends Y is weekdays
#This shows the difference in Revenue expressed in mean and other statistics
#Comparing the weekend and 
#print(Comparison.describe())


OnlyWeekdaysPro =  OnlyWeekdays.groupby(["Product_ID","DateIndex"])["Revenue"].sum()
OnlyWeekendsPro =  OnlyWeekends.groupby(["Product_ID","DateIndex"])["Revenue"].sum()
OnlyWeekdaysProMerged = pd.merge(OnlyWeekdaysPro,df_Products, on="Product_ID").drop(columns="Price")
OnlyWeekendsProMerged =  pd.merge(OnlyWeekendsPro,df_Products, on="Product_ID").drop(columns="Price")

OnlyWeekdaysProMergedFam =  pd.merge(OnlyWeekdaysPro,df_Products, how="left",on="Product_ID").drop(columns="Price").groupby("Family")["Revenue"].sum()
OnlyWeekendsProMergedFam =  pd.merge(OnlyWeekendsPro,df_Products, how="left",on="Product_ID").drop(columns="Price").groupby("Family")["Revenue"].sum()

#X is weekends Y is weekdays
#This show what Family sells more during either weekend or weekday
WeekendComparedFam = pd.merge(OnlyWeekendsProMergedFam,OnlyWeekdaysProMergedFam, on="Family")
#print(WeekendComparedFam)

OnlyWeekdaysProMergedDes =  pd.merge(OnlyWeekdaysPro,df_Products, how="left",on="Product_ID").drop(columns="Price").groupby("Description")["Revenue"].sum()
OnlyWeekendsProMergedDes =  pd.merge(OnlyWeekendsPro,df_Products, how="left",on="Product_ID").drop(columns="Price").groupby("Description")["Revenue"].sum()

#X is weekends Y is weekdays
#This show which products sell more during either weekend or weekday
WeekendComparedDis= pd.merge(OnlyWeekendsProMergedDes,OnlyWeekdaysProMergedDes, on="Description")
#print(WeekendComparedDis.head(289))

In [16]:
#########################
#Q4. What are the routes?
#########################

#How are the customers divided into regions? 
#Which customers are assigned to which routes? 
#Which routes are assigned to which depots?

In [68]:
#Let's start by merging the df_RouteTemplate datdabse with the other two databases (CustomerType and Depotfor closer detail
#The merge is done via the equally named column "HVROUTETEMPLATE_NRID" in all databases.
#However the df_RouteTemplate contains these ID's in the first column, was this was the list of keys from our dictionary
#This results in the first column having no columname to merge with other databases. To solve this we can copy this column.
#After we copied this column we can add it to our original database, we now have duplicate column with which we can work with.
df_RouteTemplate["HVROUTETEMPLATE_NRID"] = df_RouteTemplate.index

#next we will merge the two databases df_RouteTemplate and df_CustomerType in order to link our customer and route info
df_Customer_Route = pd.merge(df_CustomerType, df_RouteTemplate, how = "outer", on = "HVROUTETEMPLATE_NRID")
#for future question we will also need the depot so that's why we join them as well
df_Customer_Route_Depot = pd.merge(df_Depot, df_Customer_Route, how = "outer", on = "HVROUTETEMPLATE_NRID")

In [ ]:
#increasing the view of our dataframe, set the parameters if you want less rows to be visible
pd.set_option("display.max_rows", None, "display.max_columns", None)

In [69]:
#Q4.1 How are the customers divided into regions? 

#here we can see that every customer has its own routenumber, postcode, region, weekorder and dayofweek. 
#If we make a subselection of our total database and remove all duplicates for our region and postcode,
#we can find a database that shows us each region linked to each unique postcode, as we assumed (see picture of map)
#all unique postcodes are clustered in 4 groups en each cluster of postcodes corresponds to the same region!
#so to answer: The postcode of a customer determines its REGION
df_Customer_Route_Depot.drop_duplicates(subset = ["Postcode", "REGION"]).sort_values(by="REGION")

DEPOT  HVROUTETEMPLATE_NRID CustomerType Postcode Language  \
3468  Vilvoorde             289419955      Private     3550    nl-BE   
1279     Deinze             289414945      Private     3582    nl-BE   
3140  Vilvoorde             289416949      Private     3580    nl-BE   
1815  Antwerpen             289756470      Private     3670    nl-BE   
3401  Vilvoorde             289761480      Private     3668    nl-BE   
1406     Deinze             289757472      Private     3640    nl-BE   
1414     Deinze             289757472      Private     3960    nl-BE   
1474     Deinze             289757472      Private     3680    nl-BE   
1804  Antwerpen             289756470      Private     3660    nl-BE   
2327  Antwerpen             289763484      Private     3650    nl-BE   
0        Deinze             219020601      Private     3930    nl-BE   
1681  Antwerpen             219018687      Private     3960    nl-BE   
273      Deinze             219024456      Private     3950    nl-BE   
10       Deinze             219020601      Private     3910    nl-BE   
4257  Vilvoorde             289661606      Private     3590    nl-BE   
4526  Vilvoorde             289655594      Private     3500    nl-BE   
4598  Vilvoorde             289655594      Private     3511    nl-BE   
4728  Vilvoorde             289659602      Private     3600    nl-BE   
5680   Aarschot             289657598      Private     3501    nl-BE   

        Season REGION  WEEKORDER  DAYOFWEEK_NRID  
3468  All time    B02          2               6  
1279  All time    B02          1               6  
3140  All time    B02          2               3  
1815  All time    B25          1               3  
3401  All time    B25          2               3  
1406  All time    B25          1               4  
1414  All time    B25          1               4  
1474  All time    B25          1               4  
1804  All time    B25          1               3  
2327  All time    B25          2               5  
0     All time    C04          1               5  
1681  All time    C04          1               2  
273   All time    C04          2               6  
10    All time    C04          1               5  
4257  All time    C17          2               3  
4526  All time    C17          1               2  
4598  All time    C17          1               2  
4728  All time    C17          1               6  
5680  All time    C17          1               4

In [81]:
#Q4.2 Which customers are assigned to which routes?
df_Customer_Route_Depot.drop_duplicates(subset = ["HVROUTETEMPLATE_NRID", "WEEKORDER", "DAYOFWEEK_NRID"]
                                       ).sort_values(by=["HVROUTETEMPLATE_NRID", "WEEKORDER", "DAYOFWEEK_NRID"])

DEPOT  HVROUTETEMPLATE_NRID CustomerType Postcode Language  \
1650     Deinze              70671404     Catering     3950    nl-BE   
1651  Antwerpen             219018687      Private     3950    nl-BE   
921      Deinze             219019324      Private     3950    nl-BE   
1090     Deinze             219019962      Private     3950    nl-BE   
0        Deinze             219020601      Private     3930    nl-BE   
139      Deinze             219021241      Private     3910    nl-BE   
396      Deinze             219021882      Private     3930    nl-BE   
514      Deinze             219022524      Private     3930    nl-BE   
639      Deinze             219023167      Private     3930    nl-BE   
792      Deinze             219023811      Private     3930    nl-BE   
272      Deinze             219024456       Horeca     3910    nl-BE   
3943  Vilvoorde             289411939      Private     3582    nl-BE   
4097  Vilvoorde             289412941      Private     3550    nl-BE   
4398  Vilvoorde             289413943      Private     3582    nl-BE   
1279     Deinze             289414945      Private     3582    nl-BE   
3140  Vilvoorde             289416949      Private     3580    nl-BE   
3585  Vilvoorde             289417951      Private     3550    nl-BE   
3767  Vilvoorde             289418953      Private     3550    nl-BE   
3468  Vilvoorde             289419955      Private     3550    nl-BE   
4526  Vilvoorde             289655594      Private     3500    nl-BE   
4926  Vilvoorde             289656596      Private     3590    nl-BE   
5678   Aarschot             289657598      Private     3500    nl-BE   
5541   Aarschot             289658600      Private     3500    nl-BE   
4683  Vilvoorde             289659602      Private     3590    nl-BE   
4804  Vilvoorde             289660604      Private     3500    nl-BE   
4257  Vilvoorde             289661606      Private     3590    nl-BE   
5380   Aarschot             289662608      Private     3500    nl-BE   
5281   Aarschot             289663610      Private     3500    nl-BE   
5164  Vilvoorde             289664612      Private     3500    nl-BE   
1978  Antwerpen             289755468      Private     3680    nl-BE   
1804  Antwerpen             289756470      Private     3660    nl-BE   
1406     Deinze             289757472      Private     3640    nl-BE   
2480  Antwerpen             289758474      Private     3650    nl-BE   
2155  Antwerpen             289759476      Private     3680    nl-BE   
2974  Antwerpen             289760478      Private     3680    nl-BE   
3295  Vilvoorde             289761480      Private     3660    nl-BE   
2650  Antwerpen             289762482      Private     3680    nl-BE   
2327  Antwerpen             289763484      Private     3650    nl-BE   
2814  Antwerpen             289764486      Private     3680    nl-BE   

        Season REGION  WEEKORDER  DAYOFWEEK_NRID  
1650  All time    C04          1               1  
1651  All time    C04          1               2  
921   All time    C04          1               3  
1090  All time    C04          1               4  
0     All time    C04          1               5  
139   All time    C04          1               6  
396     Summer    C04          2               2  
514     Summer    C04          2               3  
639   All time    C04          2               4  
792   All time    C04          2               5  
272   All time    C04          2               6  
3943  All time    B02          1               3  
4097  All time    B02          1               4  
4398  All time    B02          1               5  
1279  All time    B02          1               6  
3140  All time    B02          2               3  
3585  All time    B02          2               4  
3767  All time    B02          2               5  
3468  All time    B02          2               6  
4526  All time    C17          1               2  
4926  All time    C17          1               3  
5678  A

REGION:
C04 -> 2190xxxxx
C17 -> 2896xxxxx
B02 -> 2894xxxxx
B25 -> 2897xxxxx

The other numbers of the route are less clear, however we can see a clear relationship between the ROUTENUMBER
and the cascading nature for REGION, WEEK and DAY. So based on your region, the week your delivery arives and 
the day your delivery arrives that week you get assigned a route(number).
Suppose we take an extra look and eliminate all duplicate rows, but not just for columns routenr, weekorder and dayofweek
then we can see no real connection with the routenumber and the customertype, language or season.

E.g. an example below: The routenumber 219018687 stays the same for a change in customertype, language and season. However it only changes to 219019324 when the day in the week jumps from 2 to 3 (or the depot from antwerpen to deinze), anyhow a relation towards season, language and customer type is not present. 

In [79]:
df_Customer_Route_Depot.drop_duplicates().sort_values(by=["HVROUTETEMPLATE_NRID", "WEEKORDER", "DAYOFWEEK_NRID"])

DEPOT  HVROUTETEMPLATE_NRID CustomerType Postcode Language  \
1650     Deinze              70671404     Catering     3950    nl-BE   
1651  Antwerpen             219018687      Private     3950    nl-BE   
1664  Antwerpen             219018687      Private     3950    nl-BE   
1681  Antwerpen             219018687      Private     3960    nl-BE   
1764  Antwerpen             219018687       Horeca     3950    nl-BE   
1797  Antwerpen             219018687      Private     3950    fr-BE   
921      Deinze             219019324      Private     3950    nl-BE   
953      Deinze             219019324      Private     3950    nl-BE   
1051     Deinze             219019324       Horeca     3950    nl-BE   
1083     Deinze             219019324       Horeca     3950    nl-BE   
1090     Deinze             219019962      Private     3950    nl-BE   
1109     Deinze             219019962      Private     3950    nl-BE   
1231     Deinze             219019962       Horeca     3950    nl-BE   
1232     Deinze             219019962       Horeca     3950    nl-BE   
0        Deinze             219020601      Private     3930    nl-BE   
10       Deinze             219020601      Private     3910    nl-BE   
12       Deinze             219020601      Private     3910    nl-BE   
16       Deinze             219020601      Private     3930    nl-BE   
111      Deinze             219020601       Horeca     3930    nl-BE   
139      Deinze             219021241      Private     3910    nl-BE   
140      Deinze             219021241      Private     3910    nl-BE   
396      Deinze             219021882      Private     3930    nl-BE   
397      Deinze             219021882      Private     3930    nl-BE   
509      Deinze             219021882      Private     3930    nl-BE   
514      Deinze             219022524      Private     3930    nl-BE   
515      Deinze             219022524      Private     3930    nl-BE   
593      Deinze             219022524     Catering     3930    nl-BE   
637      Deinze             219022524       Horeca     3930    nl-BE   
639      Deinze             219023167      Private     3930    nl-BE   
662      Deinze             219023167      Private     3930    nl-BE   
754      Deinze             219023167       Horeca     3930    nl-BE   
791      Deinze             219023167     Catering     3930    nl-BE   
792      Deinze             219023811      Private     3930    nl-BE   
797      Deinze             219023811      Private     3930    nl-BE   
906      Deinze             219023811     Catering     3930    nl-BE   
918      Deinze             219023811       Horeca     3950    nl-BE   
272      Deinze             219024456       Horeca     3910    nl-BE   
273      Deinze             219024456      Private     3950    nl-BE   
297      Deinze             219024456      Private     3950    nl-BE   
365      Deinze             219024456       Horeca     3950    nl-BE   
372      Deinze             219024456      Private     3950    nl-BE   
3943  Vilvoorde             289411939      Private     3582    nl-BE   
4060  Vilvoorde             289411939       Horeca     3582    nl-BE   
4066  Vilvoorde             289411939       Horeca     3580    nl-BE   
4093  Vilvoorde             289411939     Catering     3582    nl-BE   
4097  Vilvoorde             289412941      Private     3550    nl-BE   
4126  Vilvoorde             289412941       Horeca     3550    nl-BE   
4185  Vilvoorde             289412941     Catering     3550    nl-BE   
4242  Vilvoorde             289412941       Horeca     3582    nl-BE   
4398  Vilvoorde             289413943      Private     3582    nl-BE   
4504  Vilvoorde             289413943       Horeca     3582    nl-BE   
4524  Vilvoorde             289413943     Catering     3582    nl-BE   
1279     Deinze             289414945      Private     3582    nl-BE   
3140  Vilvoorde             289416949      Private     3580    nl-BE   
3142  Vilvoorde             289416949       Horeca

In [78]:
#Q4.3 Which routes are assigned to which depots?

#In the dataframe below we can see for each depot every single possible route, the total number of routes in the dataframe is
#39 which corresponds to the df_RouteTemplate dataframe, where there were 39 possible routes
#This shows that every route is linked to a certain depot

In [75]:
df_Customer_Route_Depot.drop_duplicates(subset = ["DEPOT", "HVROUTETEMPLATE_NRID"]).sort_values(by = "DEPOT")

DEPOT  HVROUTETEMPLATE_NRID CustomerType Postcode Language  \
5678   Aarschot             289657598      Private     3500    nl-BE   
5380   Aarschot             289662608      Private     3500    nl-BE   
5281   Aarschot             289663610      Private     3500    nl-BE   
5541   Aarschot             289658600      Private     3500    nl-BE   
2974  Antwerpen             289760478      Private     3680    nl-BE   
2650  Antwerpen             289762482      Private     3680    nl-BE   
2480  Antwerpen             289758474      Private     3650    nl-BE   
2327  Antwerpen             289763484      Private     3650    nl-BE   
2155  Antwerpen             289759476      Private     3680    nl-BE   
1978  Antwerpen             289755468      Private     3680    nl-BE   
1804  Antwerpen             289756470      Private     3660    nl-BE   
1651  Antwerpen             219018687      Private     3950    nl-BE   
2814  Antwerpen             289764486      Private     3680    nl-BE   
1406     Deinze             289757472      Private     3640    nl-BE   
139      Deinze             219021241      Private     3910    nl-BE   
272      Deinze             219024456       Horeca     3910    nl-BE   
1650     Deinze              70671404     Catering     3950    nl-BE   
396      Deinze             219021882      Private     3930    nl-BE   
514      Deinze             219022524      Private     3930    nl-BE   
0        Deinze             219020601      Private     3930    nl-BE   
792      Deinze             219023811      Private     3930    nl-BE   
1279     Deinze             289414945      Private     3582    nl-BE   
1090     Deinze             219019962      Private     3950    nl-BE   
921      Deinze             219019324      Private     3950    nl-BE   
639      Deinze             219023167      Private     3930    nl-BE   
5164  Vilvoorde             289664612      Private     3500    nl-BE   
4926  Vilvoorde             289656596      Private     3590    nl-BE   
4804  Vilvoorde             289660604      Private     3500    nl-BE   
4683  Vilvoorde             289659602      Private     3590    nl-BE   
4526  Vilvoorde             289655594      Private     3500    nl-BE   
4398  Vilvoorde             289413943      Private     3582    nl-BE   
4097  Vilvoorde             289412941      Private     3550    nl-BE   
3943  Vilvoorde             289411939      Private     3582    nl-BE   
3585  Vilvoorde             289417951      Private     3550    nl-BE   
3468  Vilvoorde             289419955      Private     3550    nl-BE   
3295  Vilvoorde             289761480      Private     3660    nl-BE   
3140  Vilvoorde             289416949      Private     3580    nl-BE   
4257  Vilvoorde             289661606      Private     3590    nl-BE   
3767  Vilvoorde             289418953      Private     3550    nl-BE   

        Season REGION  WEEKORDER  DAYOFWEEK_NRID  
5678  All time    C17          1               4  
5380  All time    C17          2               4  
5281  All time    C17          2               5  
5541  All time    C17          1               5  
2974  All time    B25          2               2  
2650  All time    B25          2               4  
2480  All time    B25          1               5  
2327  All time    B25          2               5  
2155  All time    B25          1               6  
1978  All time    B25          1               2  
1804  All time    B25          1               3  
1651  All time    C04          1               2  
2814  All time    B25          2               6  
1406  All time    B25          1               4  
139   All time    C04          1               6  
272   All time    C04          2               6  
1650  All time    C04          1               1  
396     Summer    C04          2               2  
514     Summer    C04          2               3  
0     All time    C04          1               5  
792   All time    C04          2               5  
1279  A

In [76]:
len(df_Customer_Route_Depot.drop_duplicates(subset = ["DEPOT", "HVROUTETEMPLATE_NRID"]).sort_values(by = "DEPOT").index)

39